### PR TITLE
GH-40400: [C++] Add support for LLD

### PR DESCRIPTION
**Pull request by cryos**
### Rationale for this change

Add support for the LLD LLVM linker, it is faster than the default linker.

### What changes are included in this PR?

Added `ARROW_USE_LLD` as a CMake option, test if GCC supports the flags and then set the linker flags if so when the option is enabled.

### Are these changes tested?

When this is enabled then the libraries and tests will use the linker flag.

### Are there any user-facing changes?

No
* GitHub Issue: #40400

### CODE REVIEW

**Comment by cryos** [comment]
@kou I think this could use some testing in the CI, is there a particular host/dockerfile that would be a good candidate? This will verify nothing breaks when it is off. I also added it to the mold if to default to mold and then use lld rather than potentially try to add both if both options were enabled.

**Review by kou — COMMENTED** [review]
> is there a particular host/dockerfile that would be a good candidate?

How about using Ubuntu 20.04 based CI jobs?

```diff
diff --git a/ci/docker/ubuntu-20.04-cpp.dockerfile b/ci/docker/ubuntu-20.04-cpp.dockerfile
index 3e3b7ac3a6..f6c7ae75f4 100644
--- a/ci/docker/ubuntu-20.04-cpp.dockerfile
+++ b/ci/docker/ubuntu-20.04-cpp.dockerfile
@@ -164,6 +164,7 @@ ENV absl_SOURCE=BUNDLED \
     ARROW_SUBSTRAIT=ON \
     ARROW_USE_ASAN=OFF \
     ARROW_USE_CCACHE=ON \
+    ARROW_USE_LLD=ON \
     ARROW_USE_UBSAN=OFF \
     ARROW_WITH_BROTLI=ON \
     ARROW_WITH_BZ2=ON \
```

We can't use mold on Ubuntu 20.04 because Ubuntu 20.04 doesn't ship mold.

**Comment by cryos** [comment]
I think this should address all of your review comments @kou and it looks more general. I tested locally, and hopefully this now gets some testing in CI.

File: cpp/cmake_modules/DefineOptions.cmake [hunk 1]
```diff
@@ -170,6 +170,8 @@ takes precedence over ccache if a storage backend is configured" ON)
 
   define_option(ARROW_USE_LD_GOLD "Use ld.gold for linking on Linux (if available)" OFF)
 
+  define_option(ARROW_USE_LLD "Use the LLVM lld for linking on Linux (if available)" OFF)
```
Comment (kou) [inline cpp/cmake_modules/DefineOptions.cmake:173]:
```suggestion
  define_option(ARROW_USE_LLD "Use the LLVM lld for linking (if available)" OFF)
```

File: cpp/cmake_modules/SetupCxxFlags.cmake [hunk 2]
```diff
@@ -656,6 +656,34 @@ if(NOT WIN32 AND NOT APPLE)
...
+    unset(LLD_LINKER_FLAGS)
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+      if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "9.1.0")
+        set(LLD_LINKER_FLAGS "-fuse-ld=lld")
+      else()
+        message(STATUS "Need GCC 9.1.0 or later to use LLD linker: ${CMAKE_CXX_COMPILER_VERSION}"
+        )
+      endif()
+    elseif(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+      set(MOLD_LINKER_FLAGS "--ld-path=${LD_LLD}")
```
Comment (kou) [inline cpp/cmake_modules/SetupCxxFlags.cmake:671]:
```suggestion
      set(LLD_LINKER_FLAGS "--ld-path=${LD_LLD}")
```

**Review by kou — COMMENTED** [review]
Could you rebase on main to fix unrelated CI failures?

**Comment by kou** [comment]
Could you also add this to use `ARROW_USE_LLD` environment variable value?

```diff
diff --git a/ci/scripts/cpp_build.sh b/ci/scripts/cpp_build.sh
index 1e09924a5e..62bb1bc13f 100755
--- a/ci/scripts/cpp_build.sh
+++ b/ci/scripts/cpp_build.sh
@@ -147,6 +147,7 @@ cmake \
   -DARROW_USE_CCACHE=${ARROW_USE_CCACHE:-ON} \
   -DARROW_USE_GLOG=${ARROW_USE_GLOG:-OFF} \
   -DARROW_USE_LD_GOLD=${ARROW_USE_LD_GOLD:-OFF} \
+  -DARROW_USE_LLD=${ARROW_USE_LLD:-OFF} \
   -DARROW_USE_MOLD=${ARROW_USE_MOLD:-OFF} \
   -DARROW_USE_PRECOMPILED_HEADERS=${ARROW_USE_PRECOMPILED_HEADERS:-OFF} \
   -DARROW_USE_STATIC_CRT=${ARROW_USE_STATIC_CRT:-OFF} \
```

**Comment by kou** [comment]
Hmm. LLD wasn't used...

https://github.com/apache/arrow/actions/runs/8569456809/job/23485397834?pr=40927#step:7:858

```text
-- Using ld linker
```

Could you apply this?

```diff
diff --git a/ci/docker/ubuntu-20.04-cpp.dockerfile b/ci/docker/ubuntu-20.04-cpp.dockerfile
index 3e3b7ac3a6..648c4832d2 100644
--- a/ci/docker/ubuntu-20.04-cpp.dockerfile
+++ b/ci/docker/ubuntu-20.04-cpp.dockerfile
@@ -101,6 +101,7 @@ RUN apt-get update -y -q && \
         libutf8proc-dev \
         libxml2-dev \
         libzstd-dev \
+        lld \
         make \
         ninja-build \
         nlohmann-json3-dev \
```

**Comment by cryos** [comment]
Thanks, still getting used to how the Arrow CI works. Done.

**Comment by cryos** [comment]
-- Using optional LLVM LLD linker

https://github.com/apache/arrow/actions/runs/8572310740/job/23494564355?pr=40927 looks like it succeeded, but then sccache timed out?

**Comment by cryos** [comment]
@kou upon a restart courtesy of @assignUser it looks like it is passing and used the LLD linker.

**Review by assignUser — APPROVED** [review]
LGTM

**Review by kou — APPROVED** [review]
+1

**Pull request merged**